pkg/apiserver/webhooks/handler.go

```diff
@@ -0,0 +1,198 @@
+	requestedAdmissionReview := v1beta1.AdmissionReview{}
+
+	// The AdmissionReview that will be returned
+	responseAdmissionReview := v1beta1.AdmissionReview{}
+
+	deserializer := codecs.UniversalDeserializer()
+	if _, _, err := deserializer.Decode(body, nil, &requestedAdmissionReview); err != nil {
+		klog.Error(err)
+		responseAdmissionReview.Response = toAdmissionResponseError(err)
+	} else {
+		if requestedAdmissionReview.Request == nil {
+			responseAdmissionReview.Response = toAdmissionResponseError(fmt.Errorf("AdmissionReview.Request is nil"))
+		} else {
+			// pass to Admitter
+			responseAdmissionReview.Response = h.a.Admit(requestedAdmissionReview)
+		}
+	}
+
+	// Return the same UID
+	if requestedAdmissionReview.Request != nil {
+		responseAdmissionReview.Response.UID = requestedAdmissionReview.Request.UID
+	}
+
+	klog.V(2).Info(fmt.Sprintf("sending response: %v", responseAdmissionReview.Response))
+
+	respBytes, err := json.Marshal(responseAdmissionReview)
+	if err != nil {
+		klog.Error(err)
+	}
+	if _, err := w.Write(respBytes); err != nil {
+		klog.Error(err)
+	}
+}
+
+func toRejectedAdmissionResponse(causes []metav1.StatusCause) *v1beta1.AdmissionResponse {
+	globalMessage := ""
+	for _, cause := range causes {
+		globalMessage = fmt.Sprintf("%s %s", globalMessage, cause.Message)
+	}
+
+	return &v1beta1.AdmissionResponse{
+		Result: &metav1.Status{
+			Message: globalMessage,
+			Code:    http.StatusUnprocessableEntity,
+			Details: &metav1.StatusDetails{
+				Causes: causes,
+			},
+		},
+	}
+}
+
+func toAdmissionResponseError(err error) *v1beta1.AdmissionResponse {
+	return &v1beta1.AdmissionResponse{
+		Result: &metav1.Status{
+			Message: err.Error(),
+			Code:    http.StatusBadRequest,
+		},
+	}
+}
+
+func allowedAdmissionResponse() *admissionv1beta1.AdmissionResponse {
+	return &admissionv1beta1.AdmissionResponse{
+		Allowed: true,
+	}
+}
+
+func validateDataVolumeResource(ar v1beta1.AdmissionReview) error {
+	resource := metav1.GroupVersionResource{
+		Group:    cdicorev1alpha1.SchemeGroupVersion.Group,
+		Version:  cdicorev1alpha1.SchemeGroupVersion.Version,
+		Resource: "datavolumes",
+	}
+	if ar.Request.Resource != resource {
+		klog.Errorf("resource is %s but request is: %s", resource, ar.Request.Resource)
+		return fmt.Errorf("expect resource to be '%s'", resource.Resource)
+	}
+	return nil
+}
+
+func toPatchResponse(original, current interface{}) *admissionv1beta1.AdmissionResponse {
+	patchType := admissionv1beta1.PatchTypeJSONPatch
+
+	ob, err := json.Marshal(original)
+	if err != nil {
+		return toAdmissionResponseError(err)
+	}
+
+	cb, err := json.Marshal(current)
+	if err != nil {
+		return toAdmissionResponseError(err)
+	}
+
+	patches, err := jsonpatch.CreatePatch(ob, cb)
+	if err != nil {
+		toAdmissionResponseError(err)
+	}
+
+	pb, err := json.Marshal(patches)
+	if err != nil {
+		return toAdmissionResponseError(err)
+	}
+
+	klog.V(3).Infof("sending patches\n%s", pb)
```

Comment:
pb is a byte[] need to cast to string before printing.

Reply:
this is fine too:  https://play.golang.org/p/4Kyup47Lmcg